aopy/utils/base.py

```diff
@@ -795,4 +795,31 @@ def multiply_mat_batch(data, mat, save_path, scale = 1, max_memory_gb = 1., dtyp
     for ibatch in range(Nbatches):
         tmp = mat @ (scale*data[ibatch*batch_size:(ibatch+1)*batch_size, :].T)
         x[ibatch*batch_size:(ibatch+1)*batch_size, :] = (tmp.T).astype(dtype)
-        x.flush()
\ No newline at end of file
+        x.flush()
+
+def generate_poisson_timestamps(mu, max_time, min_time=0., refractory_period=0.):
+    """
+    Generate timestamps following a Poisson process with mean time between events mu,
+    with a specified minimum refractory period, and that fall within a specified time window.
+
+    Args:
+        mu (float): Mean time between events in seconds.
+        max_time (float): End time of the window in seconds.
+        min_time (float, optional): Start time of the window in seconds. Default 0.
+        refractory_period (float, optional): Minimum refractory period between events in seconds. Default 0.
+
+    Returns:
+        np.ndarray: Array of timestamps within the specified time window.
+    """
+    # Initial estimate of required number of events (overestimate to ensure coverage)
+    estimated_events = int((max_time - min_time) / mu * 1.5)
+    
+    # Generate inter-event times
+    inter_event_times = np.random.exponential(mu, estimated_events)
+    inter_event_times = np.maximum(inter_event_times, refractory_period)
+    
```

Comment:
In the edge case that the user inputs a refractory period >> mu, this will cease to be a poisson distribution. Its probably worth adding a warning in this case at the beginning of the function.

Reply:
yeah i'll add a note. i think even if refractory period > 0 it is no longer exactly a poisson distribution, it gets biased by expanding some of the inter_event_times to the refractory period. but this is exactly how i set it up in BMI3D so i want to keep it that way here too.